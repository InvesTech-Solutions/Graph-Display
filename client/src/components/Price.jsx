import React, { Component } from 'react';

export default class Price extends Component {
	constructor(props) {
		super(props);
		this.subtractPrices = this.subtractPrices.bind(this)
	}

	//subtracts two dollar-cents strings and returns a result with a relevant plus or minus and dollar sign
	subtractPrices(firstPrice, secondPrice) {
		if (firstPrice === secondPrice) {
			return '+$0.00';
		}
		let dollarsFirst = parseFloat(firstPrice.slice(0, firstPrice.length - 3));
		let dollarsSecond = parseFloat(secondPrice.slice(0, secondPrice.length - 3));
		let totalCentsFirst = dollarsFirst * 100 + parseFloat(firstPrice.slice(firstPrice.length-2, firstPrice.length));
		let totalCentsSecond = dollarsSecond * 100 + parseFloat(secondPrice.slice(secondPrice.length-2, secondPrice.length));
		var total = (totalCentsFirst-totalCentsSecond)/100;
		if (total > 0) {
			return `+$${this.props.formatPrice(total)}`;
		} else {
			return `-$${this.props.formatPrice(total).substring(1)}`;
		}
	}
	render() {
		return (
			<div className = 'priceDisplay'>
				<h3>{this.props.name}</h3>
					<span id='currentPrice'>${this.props.currentPrice}</span>
				<div className = 'differences'>
					<h5 className = 'dollarsDiff'>{this.subtractPrices(this.props.currentPrice, this.props.closingPrice)}</h5>
					<h5 className = 'percentDiff'>
						({parseInt(this.props.currentPrice) > parseInt(this.props.closingPrice) ? '+' : null}
						{((this.props.currentPrice/this.props.closingPrice - 1)*100).toFixed(2)}%)
					</h5>
				</div>
			</div>
		)
	}
}