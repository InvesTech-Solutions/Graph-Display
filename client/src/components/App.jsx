import React, { Component } from 'react';
import Graph from './Graph.jsx';
import Line from './Line.jsx';
import Price from './Price.jsx';
import $ from 'jquery';


class App extends Component {
  
  constructor(props) {
    super(props);
    this.state = {
      graphData: [],
      path: '',
      line: false, //used to determine whether to display the line, circle, and date
      closest: {x: null, price: '1.00', y: null, date: null}, // used to determine where to display the line, circle and date
      date: null,
      currentCompany:null, // the company whose stock is being displayed, object with name, rating, user number, and last closing price
      currentPrice: '1.00', // the current price of the stock
      closing: '1.00', // oldest price in current timeframe
    }
    this.createPath = this.createPath.bind(this);
    this.getGraphData = this.getGraphData.bind(this);
    this.getCompanyData = this.getCompanyData.bind(this);
    this.formatPrice = this.formatPrice.bind(this);
  }

  //formats integers into price strings
  formatPrice(int){
    var str = int.toString();
    let index = str.indexOf('.');
    if(index !== str.length - 3){
        if (index === -1) {
          return `${str}.00`;
        } else {
          return `${str}0`;
        }
    }
    return `${str}`;
  }

  //creates a graph path from the price data
  createPath() {
    let path = `M${this.state.graphData[0].x},${this.state.graphData[0].y}`;
    for (let i = 1; i < this.state.graphData.length; i++) {
      path+=`L${this.state.graphData[i].x},${this.state.graphData[i].y}`
    }
    this.setState({
      path: path
    })
  }

  //keeps track of the mouse's location so that the line follows it and changes state accordingly
  onMouseMove(e) {
    const closest = {x: null, price: null, y: null, date: null}
    this.setState({x: e.clientX}, () => {
        this.setState({closest:closest}, () => {
          if(this.state.closest.x - document.getElementById('date').offsetWidth / 2 >= 0){
            document.getElementById('date').style.left = this.state.closest.x - document.getElementById('date').offsetWidth / 2 + 'px';
          }
        })
    })
    this.state.graphData.map(coords => {
      if (!closest.x || 0 <= e.clientX - closest.x < e.clientX - coords.x) {
        closest.x = coords.x;
        closest.y = coords.y;
        closest.price = coords.price;
        closest.date = coords.date;
      }
    });
  }

  // fetches company price records and formats them into graph coordinates with prices attached
  getGraphData(timeframe) {
    const tempArr = []
    let x = 0;
    $.get(`http://127.0.0.1:3000/prices/${this.state.currentCompany}/monthly`, (results) => {
      this.setState({closing: this.formatPrice(results[0].price)})
      results.forEach((datapoint) => {
        tempArr.push({x:x, price: this.formatPrice(datapoint.price), y:datapoint.price, date:datapoint['DATE_FORMAT(price_date, "%b %e %Y")']})
        x += 20
      })
      this.setState({graphData: tempArr}, () => {
        this.createPath();
      });
    });
  }

  //gets a comany's info and sets it to the currentCompany state
  getCompanyData(company) {
    $.get(`http://127.0.0.1:3000/companies/company?company=${company}`, (results) => {
      this.setState({currentCompany: results[0]},() => {
        this.setState({currentPrice: this.formatPrice(this.state.currentCompany.last_closing_price),
        });
      });
    });
  }

  componentDidMount() {
    this.getGraphData('monthly');
    this.getCompanyData('animi');
  }

  render() {
    return (
      <div className = 'mainGraphContainer'>
        <Price formatPrice = {this.formatPrice} currentPrice = {this.state.line ? this.state.closest.price : this.state.currentPrice} closingPrice = {this.state.closing}/>
        <div id ='date'>{this.state.line ? this.state.closest.date : null}</div>
        <svg onMouseMove = {this.onMouseMove.bind(this)} onMouseEnter = {() => this.setState({ line: true })} onMouseLeave = { () => this.setState({ line: false })} width = {699} height = {260} className = 'graphSVG'>
            <g transform="translate(0,260)">
            <g transform="scale(1,-1)">
              <Graph class = 'mainGraph' data = {this.state.graphData} path={this.state.path}/>
              <Line closest = {this.state.closest} show = {this.state.line} />
            </g>
            </g>
        </svg>
      </div>
    );
  }
}

export default App;
