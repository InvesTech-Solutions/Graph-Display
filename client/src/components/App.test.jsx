import React from 'react';
import App from './App.jsx';
import Price from './Price.jsx';
import { shallow } from 'enzyme';

describe('App component', () => {

	it( 'Should start with an empty graph that will get coordinates from the database', () => {
		const wrapper = shallow(<App />)
		const graphData = wrapper.state().graphData
		expect(graphData.length).toEqual(0)
	})

	it('Should have a method that formats integers and decimals into price strings', () => {
		const wrapper = shallow(<App />)
		const formatted = wrapper.instance().formatPrice(432)
		expect(formatted).toEqual('432.00')
	})

	it('Should have a method that formats datapoints within the graph height', () => {
		const wrapper = shallow(<App />)
		const formatted = wrapper.instance().formatDataPoint(120, 70, 100)
		expect(formatted).toEqual(156)
	})

	it('Should be able to create a graph path, given datapoints', () => {
		const wrapper = shallow(<App />).instance()
		wrapper.setState({graphData: [{x:0, y:1}, {x:1, y:2}]}, wrapper.createPath);
		setTimeout(() => {
			expect(wrapper.state().path).toEqual('M0,1L1,2')
		}, 100)
	})
})

describe('Price component', () => {
	it('Should have a method that can subtract two price strings and use a props method to format them as a price difference', () => {
		const wrapper = shallow(<Price formatPrice={(int)=> {
		    var str = int.toString();
		    let index = str.indexOf('.');
		    if(index !== str.length - 3) {
		        if (index === -1) {
		          return `${str}.00`;
		        } else {
		          return `${str}0`;
		        }
		    }
		    return `${str}`;
  		}} />)
		const subtracted = wrapper.instance().subtractPrices('12.32', '1.21')
		expect(subtracted).toEqual('+$11.11')
	})
})
