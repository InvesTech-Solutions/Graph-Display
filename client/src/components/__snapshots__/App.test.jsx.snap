// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link changes the class when hovered 1`] = `
<div>
  <div
    id="date"
  />
  <svg
    className="graphSVG"
    height={260}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    width={699}
  >
    <g>
      <path
        className="graphPath"
        d=""
        height={270}
        width={699}
      />
    </g>
  </svg>
</div>
`;
